core/src/input_reader/hdf5.cpp

```diff
@@ -0,0 +1,211 @@
+#include "input_reader/hdf5.h"
+
+static herr_t attr_iter(hid_t loc_id, const char *name, const H5A_info_t *info,
+                        void *ctx) {
+  std::vector<InputReader::Attribute> *attrs =
+      (std::vector<InputReader::Attribute> *)ctx;
+  attrs->emplace_back(name, info->cset, info->data_size);
+
+  return 0;
+}
+
+static herr_t obj_iter(hid_t loc_id, const char *name, const H5O_info_t *info,
+                       void *ctx) {
+  std::vector<InputReader::Object> *datasets =
+      (std::vector<InputReader::Object> *)ctx;
+
+  switch (info->type) {
+  case H5O_TYPE_GROUP:
+    break;
+  case H5O_TYPE_DATASET:
+    datasets->emplace_back(name, info->type);
+    break;
+  case H5O_TYPE_NAMED_DATATYPE:
+    LOG(WARNING) << "Unsed HDF5 type found: Datatype | " << name;
+    break;
+  default:
+    LOG(WARNING) << "Unknown HDF5 type found: " << name;
+    break;
+  }
+
+  return 0;
+}
+
+namespace InputReader {
+Object::Object(const char *name, H5O_type_t type) : name(name), type(type) {}
+
+Attribute::Attribute(const char *name, H5T_cset_t cset, hsize_t size)
+    : name(name), cset(cset), size(size) {}
+
+HDF5::HDF5(const char *file) {
+  herr_t err;
+  fileID_ = H5Fopen(file, H5F_ACC_RDONLY, H5P_DEFAULT);
+  assert(fileID_ >= 0);
+
+  err = H5Aiterate(fileID_, H5_INDEX_NAME, H5_ITER_INC, nullptr, attr_iter,
+                   (void *)&attributes_);
+  assert(err >= 0);
+  err = H5Ovisit(fileID_, H5_INDEX_NAME, H5_ITER_NATIVE, obj_iter,
+                 (void *)&datasets_, H5O_INFO_BASIC);
+  assert(err >= 0);
+
+  for (const auto &e : attributes_) {
+    LOG(DEBUG) << "InputReader - Attribute: " << e.name << " - " << e.size;
+  }
+  for (const auto &e : datasets_) {
+    LOG(DEBUG) << "InputReader - Dataset: " << e.name << " - " << e.type;
+  }
+}
+
+HDF5::~HDF5() {
+  herr_t err = H5Fclose(fileID_);
+  assert(err >= 0);
+  (void)err;
+}
+
+bool HDF5::hasAttribute(const char *name) const {
+  for (const auto &e : attributes_) {
+    if (e.name == name) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool HDF5::hasDataset(const char *name) const {
+  for (const auto &e : datasets_) {
```

Comment:
@Conni2461 Maybe prefer the use of STL function 'find_if'. The code will look more clean.